tests/test_trees.py

```diff
@@ -171,6 +173,76 @@ def integer2(cls, args):
         x = MyTransformer().transform( Tree('hello', [2]))
         self.assertEqual(x, 'hello')
 
+    def test_smart_decorator(self):
+        class OtherClass:
+            @staticmethod
+            def ab_staticmethod(a, b):
+                return (a, b)
+
+            @classmethod
+            def ab_classmethod(cls, a, b):
+                assert cls is OtherClass, cls
+                return (a, b)
+
+            def ab_method(self, a, b):
+                assert isinstance(self, OtherClass), self
+                return (a, b)
+
+        class CustomCallable:
+            def __call__(self, *args, **kwargs):
+                assert isinstance(self, CustomCallable)
+                return args
+
+        oc_instance = OtherClass()
+
+        @v_args(inline=True)
+        class TestCls(Transformer):
+            @staticmethod
+            def ab_staticmethod(a, b):
+                return (a, b)
+
+            @classmethod
+            def ab_classmethod(cls, a, b):
+                assert cls is TestCls
+                return (a, b)
+
+            def ab_method(self, a, b):
+                assert isinstance(self, TestCls)
+                return (a, b)
+
+            oc_class_ab_staticmethod = oc_instance.ab_staticmethod
+            oc_class_ab_classmethod = oc_instance.ab_classmethod
+
+            oc_ab_staticmethod = oc_instance.ab_staticmethod
+            oc_ab_classmethod = oc_instance.ab_classmethod
+            oc_ab_method = oc_instance.ab_method
+
+            set_union = set(["a"]).union
+            static_add = staticmethod(add)
+            partial_reduce_mul = partial(reduce, mul)
+
+            # custom_callable = CustomCallable()
+
+        test_instance = TestCls()
+        expected = {
+            "ab_classmethod": ([1, 2], (1, 2)),
+            "ab_staticmethod": ([1, 2], (1, 2)),
+            "ab_method": ([1, 2], (1, 2)),
+            # "oc_ab_classmethod": ([1, 2], (1, 2)),
+            # "oc_ab_staticmethod": ([1, 2], (1, 2)),
+            # "oc_class_ab_classmethod": ([1, 2], (1, 2)),
+            # "oc_class_ab_staticmethod": ([1, 2], (1, 2)),
+            # "oc_ab_method": ([1, 2], (1, 2)),
+            # "custom_callable": ([1, 2], (1, 2)),
+            "set_union": ([["b"], ["c"]], {"a", "b", "c"}),
+            "static_add": ([1, 2], 3),
+            "partial_reduce_mul": ([[1, 2]], 2),
+        }
+        for method_name, (children, expected_result) in expected.items():
+            result = test_instance.transform(Tree(method_name, children))
```

Comment:
shouldn't we aslo test the static/class methods with the uninstanciated class? you can have a perfectly functional resulting method but that didn't preserve the staticity

Reply:
Yes, will add that.